webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,760 @@
+      if (providerKey === "local") {
+        CloudDownloadsInternal.handleLocalMove(download);
+      } else {
+        CloudDownloadsInternal.selectedProvider = { providerKey, value: this.providers.get(providerKey) };
+        CloudDownloadsInternal.handleMove(download);
+      }
+      return;
+    }
+
+    switch (event.target.id) {
+    case "cloudDownloadSave":
+      if (!event.target.getAttribute("disabled")) {
+        const providerKey = event.target.getAttribute("providerKey");
+        CloudStorage.savePromptResponse(providerKey, true, true);
+        CloudDownloadsInternal.selectedProvider = { providerKey, value: this.providers.get(providerKey) };
+        // Prepare for future downloads move to Download Folder by checking
+        // if Download folder exists in cloud provider folder, if not create one
+        CloudDownloadsInternal.checkProviderDownloadFolder();
+        event.currentTarget.setAttribute("hidden", "true");
+      }
+      break;
+    case "cloudDownloadCancel":
+      // Set interval when notification was last shown
+      Services.prefs.setIntPref(CLOUD_SERVICES_PREF + "lastprompt",
+        Math.floor(Date.now() / 1000));
+      event.currentTarget.setAttribute("hidden", "true");
+      break;
+    case "cloudDownloadPreference": {
+      const origin = null;
+      const entryPoint = "CloudStorage";
+      this.getRecentWindow().openPreferences("paneGeneral", {origin, urlParams: {entrypoint: entryPoint}});
+      break;
+    }
+    }
+  },
+};
+
+
+// Cloud Downloads Internal API that observe downloads and handle
+// downloaded item move once user selects Move to provider folder context menu option
+var CloudDownloadsInternal = {
+  /**
+   * Provider selected in context menu
+   */
+  selectedProvider: null,
+
+  /**
+    * Icon URL shown before default Local Download Dir in context menu
+    */
+  defaultDownloadDirIconURL: null,
+
+  /**
+   * Checks if the asset with input path exist on
+   * file system
+   * @return {Promise}
+   * @resolves
+   * boolean value of file existence check
+   */
+  checkIfAssetExists(path) {
+    return OS.File.exists(path).catch(err => {
+      Cu.reportError(`Couldn't check existance of ${path}`, err);
+      return false;
+    });
+  },
+
+  checkIfExistingCloudProviderDownloadSettings() {
+    // Check if user has previously set preferred download directoy path as
+    // one of cloud provider folder, if yes exit without showing notification
+    // Telemetry - how many such users
+    if (!CloudDownloadsView.providers) {
+      return false;
+    }
+    const dwnldDirPath = this.downloadDirSetting; // await Downloads.getPreferredDownloadsDirectory();
+    const providerValues = [...CloudDownloadsView.providers.values()];
+    return providerValues.some(v => dwnldDirPath.includes(v.downloadPath));
+  },
+
+  async _setDefaultDownloadDirIconURL() {
+    const downloadsDir = await Downloads.getPreferredDownloadsDirectory();
```

Comment:
This can also return the cloud one, right? ( https://dxr.mozilla.org/mozilla-central/rev/63519bfd42ee379f597c0357af2e712ec3cd9f50/toolkit/components/downloads/DownloadIntegration.jsm#309-316 )

But then we use this as the icon for the "move to local" menuitem. That seems wrong...

Reply:
For user who has previously opted-in by selecting cloud as default download (has cloud.services.storage.key set), we are not displaying context menu.
https://github.com/mozilla/shield-cloudstorage/blob/cloud-storage-web-ext/webextension/privileged/ui/CloudDownloadsView.jsm#L451